doc/specs/stdlib_hash_functions.md

```diff
@@ -0,0 +1,1697 @@
+    successors. We intend this dedication to be an overt act of
+    relinquishment in perpetuity of all present and future rights to this
+    software under copyright law.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+    IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+    OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+    ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+    OTHER DEALINGS IN THE SOFTWARE.
+
+    For more information, please refer to <http://unlicense.org>
+
+`PENGY_HASH` is a translation to Fortran 2008 and signed two's
+complement arithmetic of the `pengyhash` algorithm of Alberto Fajardo,
+copyright 2020. Alberto Fajardo's original C code, `pengyhash.c`, is
+available at the URL:
+https://github.com/tinypeng/pengyhash/blob/master/pengyhash.c
+under the BSD 2-Clause License:
+https://github.com/tinypeng/pengyhash/blob/master/LICENSE
+
+The BSD 2-Clause license is as follows:
+
+    BSD 2-Clause License
+
+    pengyhash
+    Copyright (c) 2020 Alberto Fajardo
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions
+    are met:
+
+    1. Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above
+    copyright notice, this list of conditions and the following
+    disclaimer in the documentation and/or other materials provided
+    with the distribution.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+	CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+	INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+	MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+	DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS
+	BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+	EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+	DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+	ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+	TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+	THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+	SUCH DAMAGE.
+
+
+## The hash codes modules
+
+### Overview of the modules
+
+The Standard Library provides two modules implementing hash
+functions and scalar hashes.
+The `stdlib_32_bit_hash_functions` module provides procedures to
+compute 32 bit integer hash codes and a scalar hash. 
+The hash codes are useful for tables of up to `2**15` entries, and
+for keys with a few hundred elements.
```

Comment:
I don't quite understand this. I think the first part means `useful for indexing hash tables with up to 2**15 entries`? But I'm not sure what `keys` are -- is this the object for which we compute a hash (so an array of integers or a character string)? Likely some slight rewording will make that clearer.

Reply:
This needs to be changed. Keys are what is hashed. The true identifier of the data. I have been testing my hash tables with these codes and they seem to be working well of `2**16` random elements. Let me update the documentation after a little more feedback.

Reply:
I have changed the discussion to try to clarify the issues.